src/Symfony/Component/Messenger/Transport/Enhancers/MemoryLimitReceiver.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport\Enhancers;
+
+use Symfony\Component\Messenger\Transport\ReceiverInterface;
+
+/**
+ * @author Simon Delicata <simon.delicata@free.fr>
+ */
+class MemoryLimitReceiver implements ReceiverInterface
+{
+    private $decoratedReceiver;
+    private $memoryLimit;
+    private $memoryResolver;
+
+    public function __construct(ReceiverInterface $decoratedReceiver, string $memoryLimit, callable $memoryResolver = null)
+    {
+        $this->decoratedReceiver = $decoratedReceiver;
+        $this->memoryLimit = $this->convertToOctets($memoryLimit);
+        $this->memoryResolver = $memoryResolver ?? function () {
+            return \memory_get_usage();
+        };
+    }
+
+    public function receive(callable $handler): void
+    {
+        $this->decoratedReceiver->receive(function ($message) use ($handler) {
+            $handler($message);
+
+            $memoryResolver = $this->memoryResolver;
+            if ($memoryResolver() >= $this->memoryLimit) {
+                $this->stop();
+            }
+        });
+    }
+
+    public function stop(): void
+    {
+        $this->decoratedReceiver->stop();
```

Comment:
I wondering if should we throw some kind of exception or log after that? I mean, right now it's stopped silently, how can we differentiate if it was finished by memory limit or because it has received all messages?

Reply:
@yceruto Good idea.
@sroze Did you already imagined such a notification mechanism?

Reply:
I don't see any reason throwing an exception for this. What is your use-case? Do you need to know this for something?

Reply:
Actually, it is probably a good idea to log such behavior (no exception). Maybe an optional `LoggerInterface` argument that would be called with `info` when that happens? (note that it wouldn't be on this line but when you actually detect the memory usage line 41 of this line).

@lyrixx any PoV on the logging level?

Reply:
I disagree exception too, but the optional LoggerInterface is a good idea